syntax = "proto3";

package pbft;

message Block {
    string previous_blockhash = 1;
    string blockhash = 2;
    int32 blockheight = 3;
}

message PBFTMessage {
    string sender_id = 1;
    Block block = 2;
    string phase = 3; // PrePrepare, Prepare, Commit
    bytes signature = 4; 
}
message PublicKeyMessage {
    string sender_id = 1;
    bytes public_key = 2;
}

message Empty {}
message ViewChangeMessage {
    int32 view = 1;       // Số view mới
    string node_id = 2;   // ID của node gửi thông điệp
}

message NewViewMessage {
    int32 view = 1;       // Số view mới
    string node_id = 2;   // ID của node trở thành primary
}


service PBFTService {
    rpc PrePrepare(PBFTMessage) returns (Empty);
    rpc Prepare(PBFTMessage) returns (Empty);
    rpc Commit(PBFTMessage) returns (Empty);
    rpc ViewChange(ViewChangeMessage) returns (Empty);  // Thêm dịch vụ ViewChange
    rpc NewView(NewViewMessage) returns (Empty);        // Thêm dịch vụ NewView
    rpc ExchangePublicKey(PublicKeyMessage) returns (Empty);  // Dịch vụ trao đổi khóa công khai
}